[mlir][memref] Transpose: allow affine map layouts in result, extend folder

ubfx opened the pull request:
Currently, the `memref.transpose` verifier forces the result type of the Op to have an explicit `StridedLayoutAttr` via the method `inferTransposeResultType`. This means that things like the example Op given in the documentation (https://mlir.llvm.org/docs/Dialects/MemRef/#memreftranspose-memreftransposeop) is actually invalid because it uses an `AffineMap` to specify the layout:
```mlir
%1 = memref.transpose %0 (i, j) -> (j, i) : memref<?x?xf32> to memref<?x?xf32, affine_map<(d0, d1)[s0] -> (d1 * s0 + d0)>>
```

It also means that we can't "un-transpose" a transposed memref back to the implicit layout form, because the verifier will always enforce the explicit strided layout.

This patch makes the following changes:
1. `inferTransposeResultType()` returns a `MemRefType` with canonicalized strided layout, i.e the strides are turned into a linearizing affine expression.
2. The verifier checks whether the canonicalized strided layout of the result Type is identitcal to the infered (also canonical) result type layout. This way, it's only important that the two Types have the same strided layout, not necessarily the same representation of it.
3. The folder is extended to support folding away the trivial case of identity permutation and to fold one transposition into another by composing the permutation maps.

ubfx committed:
[mlir][memref] Transpose: allow affine map layouts in result, extend folder

ubfx committed:
newline

ubfx committed:
compose the right way around and add a test that would have caught the issue

lipracer reviewed (approved):
LGTM. Please be patient and wait for other reviewers to have any other comments.

ubfx committed:
Infer un-canonicalized strided layout

ubfx commented:
I changed the patch so that `inferTransposedResultType()` returns a `MemRefType` with `StridedLayoutAttr`, which is the same as it was before this patch. I moved the change to the verifier, where now, we just make sure that the canonicalized Version of the Op's result Ttpe equals the canonicalized version of the inferred result type. This way, all existing code that creates `memref::TransposeOp` with infered result Type should behave the same after the patch, but we still allow the Op to be used (and folded) with Affine Map layouts.

ftynse reviewed (approved):
I think this was just overlooked when we transitioned to the first-class support for strided layouts and the intention would have been to have the map result type have the strided layout if the input had.
